### PR TITLE
Add initial ies lights.

### LINKED ISSUE #715: IES Lights using the light projector system
**Describe the project you are working on:**
Not related to my project.

**Describe the problem or limitation you are having in your project:**
Lights in Godot Engine are clunky and look the same.

There are existing IES profiles that give varied and exciting light types.

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
It would make artwork and level design improved if IES light profiles can be imported directly.

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**
1. Read IES profile
2. Generate a format the light projector feature can use
3. Import the format as a light projector

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**

No, the convenience of dragging a profile to be a light project is the feature.

**Is there a reason why this should be core and not an add-on in the asset library?:**
In theory this can be a gdnative plugin, but ies profiles is a important visual tool in lighting a scene.

### PR DESCRIPTION
**Pull request by fire**
Fixes: https://github.com/godotengine/godot-proposals/issues/715

Status:

- [x] Copyright handling

Imports something into an image. Not sure how projectors work.

image

Part of the V-Sekai project. https://github.com/v-sekai

### CODE REVIEW

**Comment by jcostello** [comment]
How does it works?

**Comment by fire** [comment]
I currently have generated a 521x1 strip.

My next task is to convert that to a cube map.

```
float interpolate2D(const IESFileInfo& info, float angleV, float angleH) const;
```

Assistance helpful.

**Comment by fire** [comment]
@clayjohn Said he'll describe how to integrate with the light projectors.

```hlsl
// From the original guide
// hlsl
float3 ComputeIESLightProfile1D(float3 L, float3 LightDirection)
{
    L = mul(L, (float3x3)matViewInverse);
    float angle = (1 - acos(dot(L, LightDirection)) / PI);
    return tex2Dlod(IESTextureSamp, float4(angle, 0, 0, 0)).rgb;
}
```

File: modules/gridmap/grid_map.cpp [hunk 1]
```diff
@@ -572,11 +572,12 @@ bool GridMap::_octant_update(const OctantKey &p_key) {
 				NavigationServer3D::get_singleton()->region_set_map(region, get_world_3d()->get_navigation_map());
 				nm.region = region;
 
+#ifdef DEBUG_ENABLED
 				// add navigation debugmesh visual instances if debug is enabled
 				SceneTree *st = SceneTree::get_singleton();
 				if (st && st->is_debugging_navigation_hint()) {
 					if (!nm.navmesh_debug_instance.is_valid()) {
-						RID navmesh_debug_rid = navmesh->get_debug_mesh()->get_rid();
+						RID navmesh_debug_rid = navmesh->_get_debug_mesh()->get_rid();
```
Comment (reduz) [inline modules/gridmap/grid_map.cpp:580]: seems unrelated?

**Review by reduz — CHANGES_REQUESTED** [review]

**Comment by reduz** [comment]
This looks good, why a draft?

File: modules/ies/image_loader_ies.cpp [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#include "image_loader_ies.h"
+
+#include "core/io/file_access.h"
+#include "core/string/ustring.h"
+
+#include "thirdparty/ies/ies_loader.h"
+#include <ios>
```
Comment (akien-mga) [inline modules/ies/image_loader_ies.cpp:37]: What needs this? I only see `std::string`.

File: thirdparty/README.md [hunk 3]
```diff
@@ -247,6 +247,15 @@ Files generated from upstream source:
 - Step 4: Copy `source/data/out/icudt71l.dat` to the `{GODOT_SOURCE}/thirdparty/icu4c/icudt71l.dat`.
 
 
+## ies
+
+- Upstream: https://github.com/ray-cast/IES
+- Version: master (d6d36e5fd648049ed4ade087d5427896a54c1a26, 2018)
```
Comment (akien-mga) [inline thirdparty/README.md:253]:
```suggestion
- Version: git (d6d36e5fd648049ed4ade087d5427896a54c1a26, 2018)
```

File: thirdparty/README.md [hunk 4]
```diff
@@ -247,6 +247,15 @@ Files generated from upstream source:
...
+- License: MIT
+
+See the standard from https://github.com/ray-cast/ies/blob/master/IES.pdf and from http://www.tc.faa.gov/its/worldpac/Standards/ansi/ansi-ies%20rp-1-04.pdf
```
Comment (akien-mga) [inline thirdparty/README.md:256]:
```suggestion
See the standard from https://github.com/ray-cast/ies/blob/master/IES.pdf and from
http://www.tc.faa.gov/its/worldpac/Standards/ansi/ansi-ies%20rp-1-04.pdf.
```

File: thirdparty/ies/ies_loader.h [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+private:
+	static void skipSpaceAndLineEnd(const std::string& data, std::string& out, bool stopOnComma = false);
+
+	static void getLineContent(const std::string& data, std::string& next, std::string& line, bool stopOnWhiteSpace, bool stopOnComma);
+	static void getFloat(const std::string& data, std::string& next, float& ret, bool stopOnWhiteSpace = true, bool stopOnComma = false);
+	static void getInt(const std::string& data, std::string& next, std::int32_t& ret, bool stopOnWhiteSpace = true, bool stopOnComma = false);
+};
+
+#endif
```
Comment (akien-mga) [inline thirdparty/ies/ies_loader.h:116]: Missing newline at EOF.

File: thirdparty/ies/ies_loader.cpp [hunk 6]
```diff
@@ -0,0 +1,582 @@
...
+}
+
+void
+IESLoadHelper::getInt(const std::string& data, std::string& next, std::int32_t& ret, bool stopOnWhiteSpace, bool stopOnComma)
+{
+	std::string line;
+	getLineContent(data, next, line, stopOnWhiteSpace, stopOnComma);
+	assert(!line.empty());
+	ret = std::atoi(line.c_str());
+}
```
Comment (akien-mga) [inline thirdparty/ies/ies_loader.cpp:582]: Missing newline.

File: thirdparty/README.md (hunk 4 above)
Comment (akien-mga) [inline thirdparty/README.md:256]:
```suggestion
- License: MIT

Files extracted from upstream source:

- `ies_loader.{cpp,h}`
- `LICENSE.txt`

See the standard from https://github.com/ray-cast/ies/blob/master/IES.pdf and from http://www.tc.faa.gov/its/worldpac/Standards/ansi/ansi-ies%20rp-1-04.pdf
```

File: COPYRIGHT.txt [hunk 7]
```diff
@@ -236,6 +236,11 @@ Comment: International Components for Unicode
 Copyright: 1991-2021, Unicode
 License: Unicode
 
+Files: ./thirdparty/ies/
+Comment: ies
```
Comment (akien-mga) [inline COPYRIGHT.txt:240]:
```suggestion
Comment: IES Light Profiles
```

**Review by akien-mga — CHANGES_REQUESTED** [review]

**Comment by fire** [comment]
@reduz I wasn't sure how to take the 1x512 px image and turn it into a projection that Godot can use for cookie lighting.

**Comment by Calinou** [comment]
> I wasn't sure how to take the 1x512 px image and turn it into a projection that Godot can use for cookie lighting.

OmniLights use a texture with spherical projection (like HDRI panoramas, but [upside down](https://github.com/godotengine/godot/issues/57704**)). SpotLights use box texture projection in the direction the SpotLight is pointing towards.

You should be able to correct the OmniLight projector's perspective using a polar to cartesian (or cartesian to polar) function.

**Comment by jcostello** [comment]
Please ensure that this feature works with static lights when baking lightmaps

**Comment by fire** [comment]
@jcostello I do not know if projectors bake. The goal of this pr is to generate a image that can be loaded by the light cookie projector.

**Comment by jcostello** [comment]
@fire I will test it once this is merged

File: modules/gridmap/grid_map.cpp (hunk 1 above)
Comment (akien-mga) [inline modules/gridmap/grid_map.cpp:580]: Unrelated, should be removed (you can open an issue about what that bug is because it's weird you keep changing it - but I suspect you might be working off a slightly outdated `master` branch if this breaks).

Comment (fire) [reply modules/gridmap/grid_map.cpp:580]: This happens on all my rebases up to last night. I also have the same error on ming-llvm and osx's clang.

**Comment by fire** [comment]
This is salvageable. Will need to re-evaluate in 4.x

**Comment by Nikoraito** [comment]
aaaa bring it back!!
